core/src/types/chain.rs

```diff
@@ -16,6 +17,159 @@ pub const CHAIN_ID_PREFIX_MAX_LEN: usize = 19;
 /// Separator between chain ID prefix and the generated hash
 pub const CHAIN_ID_PREFIX_SEP: char = '.';
 
+/// Configuration parameter for the upper limit on the number
+/// of bytes transactions can occupy in a block proposal.
+#[derive(
+    Copy,
+    Clone,
+    Eq,
+    PartialEq,
+    Ord,
+    PartialOrd,
+    Hash,
+    Debug,
+    BorshSerialize,
+    BorshDeserialize,
+)]
+pub struct ProposalBytes {
+    inner: NonZeroU64,
+}
+
+impl Serialize for ProposalBytes {
+    fn serialize<S>(&self, s: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        s.serialize_u64(self.inner.get())
+    }
+}
+
+impl<'de> Deserialize<'de> for ProposalBytes {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        struct Visitor;
+
+        impl<'de> serde::de::Visitor<'de> for Visitor {
+            type Value = ProposalBytes;
+
+            fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
+                write!(f, "a u64 in the range 1 - 94371840")
+            }
+
+            fn visit_u64<E>(self, size: u64) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                ProposalBytes::new(size).ok_or_else(|| {
+                    serde::de::Error::invalid_value(
+                        serde::de::Unexpected::Unsigned(size),
+                        &self,
+                    )
+                })
+            }
+
+            fn visit_i64<E>(self, size: i64) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                ProposalBytes::new(size as u64).ok_or_else(|| {
+                    serde::de::Error::invalid_value(
+                        serde::de::Unexpected::Signed(size),
+                        &self,
+                    )
+                })
+            }
```

Comment:
Do we need this `visit_i64`, would we ever expect it to get called? `inner` should always be de/serialized as a u64?

Reply:
Crappy problem with the toml library.

Reply:
deserialization fails without an `i64` visitor 😓